pennylane/gradients/finite_difference.py

```diff
@@ -0,0 +1,278 @@
+    >>> finite_diff_coeffs(n=2, approx=2, strategy="center")
+    array([[-2.,  1.,  1.],
+           [ 0., -1.,  1.]])
+    """
+    if n < 1 or not isinstance(n, int):
+        raise ValueError("Derivative order n must be a positive integer.")
+
+    if approx < 1 or not isinstance(approx, int):
+        raise ValueError("Approximation order must be a positive integer.")
+
+    num_points = approx + 2 * np.floor((n + 1) / 2) - 1
+    N = num_points + 1 if n % 2 == 0 else num_points
+
+    if strategy == "forward":
+        shifts = np.arange(N, dtype=np.float64)
+
+    elif strategy == "backward":
+        shifts = np.arange(-N + 1, 1, dtype=np.float64)
+
+    elif strategy == "center":
+        if approx % 2 != 0:
+            raise ValueError("Centered finite-difference requires an even order approximation.")
+
+        N = num_points // 2
+        shifts = np.arange(-N, N + 1, dtype=np.float64)
+
+    else:
+        raise ValueError(
+            f"Unknown strategy {strategy}. Must be one of 'forward', 'backward', 'center'."
+        )
+
+    # solve for the coefficients
+    A = shifts ** np.arange(len(shifts)).reshape(-1, 1)
+    b = np.zeros_like(shifts)
+    b[n] = factorial(n)
+    coeffs = np.linalg.solve(A, b)
+
+    coeffs_and_shifts = np.stack([coeffs, shifts])
+
+    # remove all small coefficients and shifts
+    coeffs_and_shifts[np.abs(coeffs_and_shifts) < 1e-10] = 0
+
+    # remove columns where the coefficients are 0
+    coeffs_and_shifts = coeffs_and_shifts[:, ~np.all(coeffs_and_shifts == 0, axis=0)]
+
+    # sort columns in ascending order according to abs(shift)
+    coeffs_and_shifts = coeffs_and_shifts[:, np.argsort(np.abs(coeffs_and_shifts)[1])]
+    return coeffs_and_shifts
+
+
+def generate_shifted_tapes(tape, idx, shifts, multipliers=None):
+    r"""Generate a list of tapes where the corresponding trainable parameter
+    index has been shifted by the values given.
+
+    Args:
+        tape (.QuantumTape): input quantum tape
+        idx (int): trainable parameter index to shift the parameter of
+        shifts (Sequence[float or int]): sequence of shift values
+        multipliers (Sequence[float or int]): Sequence of multiplier values to
+            scale the parameter by. If not provided, the parameter will
+            not be scaled.
+
+    Returns:
+        list[QuantumTape]: List of quantum tapes. Each tape has parameter
+        ``idx`` shifted by consecutive values of ``shift``. The length
+        of the returned list of tapes will match the length of ``shifts``.
+    """
+    params = tape.get_parameters()
+    tapes = []
+
+    for i, s in enumerate(shifts):
+        new_params = params.copy()
+        shifted_tape = tape.copy(copy_operations=True)
+
+        if multipliers is not None:
+            m = multipliers[i]
+            new_params[idx] = new_params[idx] * qml.math.convert_like(m, new_params[idx])
+
+        new_params[idx] = new_params[idx] + qml.math.convert_like(s, new_params[idx])

```

Comment:
I like that much more than the array computation!

Reply:
same 😆